crates/primitives/src/transaction/signature.rs

```diff
@@ -17,41 +17,34 @@ pub struct Signature {
 }
 
 impl Signature {
-    /// Encode the `v`, `r`, `s` values without a RLP header.
-    /// Encodes the `v` value using the legacy scheme without EIP-155.
-    pub(crate) fn encode_inner_legacy(&self, out: &mut dyn reth_rlp::BufMut) {
-        (self.odd_y_parity as u8 + 27).encode(out);
-        self.r.encode(out);
-        self.s.encode(out);
-    }
-
     /// Output the length of the signature without the length of the RLP header, using the legacy
-    /// scheme without EIP-155.
-    pub(crate) fn payload_len_legacy(&self) -> usize {
-        (self.odd_y_parity as u8 + 27).length() + self.r.length() + self.s.length()
+    /// scheme with EIP-155 support depends on chain_id.
+    pub(crate) fn payload_len_legacy(&self, chain_id: Option<u64>) -> usize {
+        self.v(chain_id).length() + self.r.length() + self.s.length()
     }
 
     /// Encode the `v`, `r`, `s` values without a RLP header.
-    /// Encodes the `v` value with EIP-155 support, using the specified chain ID.
-    pub(crate) fn encode_eip155_inner(&self, out: &mut dyn reth_rlp::BufMut, chain_id: u64) {
-        // EIP-155: v = {0, 1} + CHAIN_ID * 2 + 35
-        let v = chain_id * 2 + 35 + self.odd_y_parity as u64;
-        v.encode(out);
+    /// Encodes the `v` value using the legacy scheme with EIP-155 support depends on chain_id.
+    pub(crate) fn encode_legacy(&self, out: &mut dyn reth_rlp::BufMut, chain_id: Option<u64>) {
```

Comment:
nit: the original reason for calling these `legacy` was to indicate that the methods did not use EIP-155. So because these methods now distinguish EIP-155 based on the `Option`, we probably don't need to call these `encode_legacy` or `payload_len_legacy`

Reply:
We have three kinds of `Transaction`, Legacy, EIP2930, and EIP1559. In terms of `Signature`, `Legacy` transaction could be EIP155 or not. So, I intend and need at here to use `xxx_legacy()` methods for `Legacy` transaction and also `xxx()` methods for `EIP2930` or `EIP1559`.

Reply:
Please note that my PR description especially:

* The transaction w/ eip155 is still a legacy transaction. So I'd like to unify the interface as encode_legacy() and decode_legacy().
* For non legacy transaction, the signature encoding is processed in Transaction. For consistency, Signature should also have encode() and decode() for non legacy transaction.

Reply:
I find the name `legacy` here very confusing tbh.
how about `encode_with_eip155_chain_id` or something, because "legacy" applies to the type tx, which may or may not have a chain id

Reply:
Ah, you're right. I was mistaken and forgot that EIP-155 is only relevant for legacy transactions. Thank you for the explanation! I am OK with the method names and interface now